### PR TITLE
reverse proxy refactor

### PR DESCRIPTION
**Pull request by bcwaldon**
This is still light on tests, but I wanted to get it up for review sooner rather than later.

### CODE REVIEW

File: pkg/gateway/kubernetes.go [hunk 1]
```diff
@@ -11,21 +11,23 @@ import (
 	"strings"
 )
 
-type ServiceMapperConfig struct {
+type kubernetesReverseProxyConfigGetterConfig struct {
```
Comment (bcwaldon) [inline pkg/gateway/kubernetes.go]: naming help appreciated here

Comment (jacobstr) [inline pkg/gateway/kubernetes.go]: Lets ask a Senior Java Solution Architect.

Comment (jacobstr) [inline pkg/gateway/kubernetes.go]: I feel like this annotation prefix should likely be the same regardless of where it's coming from in Kube. Could it just be part of a KubernetesConfig?

File: pkg/gateway/kubernetes.go [hunk 2]
```diff
@@ -73,121 +75,143 @@ func getKubernetesClientConfig(kubeconfig string) (*krestclient.Config, error) {
...
 			continue
 		} else {
-			svc.TargetPort = port.TargetPort.IntValue()
-			break
+			return port.TargetPort.IntValue(), nil
 		}
 	}
 
-	return nil
+	return 0, fmt.Errorf("coult not find port matching %d for service %s in namespace %s", svcPort, svcName, svcNamespace)
```
Comment (jacobstr) [inline pkg/gateway/kubernetes.go]: Why `0` instead of `nil`?

**Comment by jacobstr** [comment]
👍 Looks badass.

File: pkg/gateway/kubernetes.go (hunk 2 above)
Comment (bcwaldon) [inline pkg/gateway/kubernetes.go]: The return types are `int, error`. `nil` is not a valid value for an `int`

File: pkg/gateway/kubernetes.go (hunk 1 above)
Comment (bcwaldon) [inline pkg/gateway/kubernetes.go]: Yes, it totally could. I think some refactoring will fall out here as we move to writing data into the annotations.

**Pull request merged**